text/0000-contract-metadata.md

```diff
@@ -0,0 +1,117 @@
+- Proposal Name: `contract_metadata`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- NEP PR: [nearprotocol/neps#0003](https://github.com/nearprotocol/NEPs/pull/3)
+
+# Summary
+[summary]: #summary
+
+This NEP introduces contract metadata, which summarizes the content of a given contract in json format.
+Contract metadata allows developers to easily list the methods in a contract and potentially display them to users to
+provide transparency.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there is no direct way for a developer to programmatically list the methods of a contract that is deployed on chain
+because the contract code stored on chain is the compiled wasm code, where the methods name and parameters are already mangled.
+As a result, if developers want to display the methods of a contract to an user of the app to provide transparency,
+especially in the case of financial apps, they have no choice but to hardcode them in the front end, which is suboptimal in many ways.
+Furthermore, if developers want to get the list of methods in some contract for some downstream task like data analysis,
+they have no way of doing so. Contract metadata aims to solve the aforementioned problems by providing a convenient way 
+for developers to list the methods of contracts.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For developers, there will be two main changes:
+- Instead of annotating view and change methods in the `initContract` function in `main.js`,
+they will instead annotate the methods of a contract by decorators.
+More specifically, every method is by default a change method, unless annotated by `@view_method`.
+- Every contract will have a `metadata` method that returns a json that serializes the contract methods in an array.
+For each method, the json serialization is of the form `{"name": <name>, "parameters": [{"name": <param_name1>, "type": <param_type1>, ..}, .. ], "returnType": <return_type>}`.
+
+As an concrete example, suppose we have a contract that maintains a counter on chain:
+
+```typescript
+import { context, storage, near } from "./near";
+
+export function incrementCounterBy(amount: i32 = 1): void {
+  let newCounter = storage.get<i32>("counter") + amount;
+  storage.set<i32>("counter", newCounter)
+  near.log("Counter is now: " + newCounter.toString());
+}
+
+export function decrementCounterBy(amount: i32 = 1): void {
+  let newCounter = storage.get<i32>("counter") - amount;
+  storage.set<i32>("counter", newCounter)
+  near.log("Counter is now: " + newCounter.toString());
+}
+
+@view_method
+export function getCounter(): i32 {
+  return storage.get<i32>("counter");
+}
+```
+
+This contract has two change methods, `incrementCounter` and `decrementCounter`, as well as one view method, `getCounter`.
+In this case, the metadata we want looks like 
+```json
+[
```

Comment:
This looks good but kinda still seems not diverse enough. E.g. it is not clear how do we represent nested objects, arrays, strings, uint128, etc.
Maybe we need few examples to cover it doesn't have to be single contract.